### ISSUE TITLE
ViT-Tiny: forward + training + timm checkpoint loader (blocks Tao E1)

### ISSUE THREAD

**Issue by OriPekelman** [OriPekelman/toy#13]
**Blocks**: `specs/E1-vit-pure-emb-n3.md` in the Tao sibling — the cheap vision
headline (granite_transfer exp #28: ViT-Tiny pure-embedding transfer, IN-21k
AugReg → Flowers-102 & Stanford Cars, 3 arms × 3 seeds; minutes per seed on the
GB10).

### What's needed

1. **ViT-Tiny target arch** in Toy: patch embedding (Conv2d 3→192, kernel 16,
   stride 16), learned positional embedding, class token, 12 transformer blocks
   (LayerNorm + MSA + MLP, GELU), final norm + MLP head.
2. **Weight loader** importing `vit_tiny_patch16_224` (timm IN-21k AugReg) into
   the matching ggml/gguf tensor layout. For E1's pure-embedding arm only
   `patch_embed`, `pos_embed`, and `cls_token` need to round-trip; block weights
   are randomly initialised in every arm.
3. **Image-classification training loop** with cross-entropy + top-1 tracking,
   emitted via `eval` events (`name:"validation"`, `extra:{ top1: 0.59 }`) —
   `events.jsonl` per the frozen `toy/v1` schema (`docs/events-schema.md`).

### Acceptance

`examples/07_train_vit_tiny.rb` runs a 200-step smoke on a tiny CIFAR-like
loader without crashing, emits well-formed `events.jsonl`, and final loss < init
loss. (Then E1 reproduces the granite_transfer #28 pure-emb result through Tao.)

### Out of scope

DeiT, ViT-S/B/L, multi-scale wavelet inits — follow-ups if E1 confirms the
recipe is worth pursuing in Toy.

---
_Filed from the Tao sibling's cross-project review — see `tao/ISSUES-FOR-TOY.md`. Schema contract: `toy/docs/events-schema.md` (frozen v1)._

**Comment by OriPekelman** [comment]
**Scope-and-decompose pass (2026-05-27).** Wrote `docs/roadmap/e1-e2-scope-2026-05-27.md` with the full E1 / ViT-Tiny breakdown — 6 sub-issues (Conv2D/Im2Col FFI → patch_embed → ViT forward graph → timm extractor → image loader → train entry point), realistic effort 5–8 days.

**Cheapest first step (E1.1).** Wrap `ggml_im2col` / `ggml_im2col_back` / `ggml_conv_2d` in TinyNN FFI. Pure C-side binding, no Ruby graph changes, single-PR-sized.

Holding off on splitting this into individual GitHub issues until we actually start the work — they live in the doc as a checklist for now to keep the tracker focused on landed/active work.

**Comment by OriPekelman** [comment]
**E1.1 landed** (commit `a404d6d`). Conv2D + Im2Col primitives bound; PyTorch parity at 0.000e+00 max abs diff on a 4×4×3 input with 2×2×3×2 kernel. Building blocks for E1.2 (patch_embed composite) are in place — that's the next sub-issue per the decomp doc.

PyTorch is now declared as an inline-script dep (PEP 723) in `tinynn/conv2d_parity.py`; `uv run` handles the venv transparently. Same pattern as `prep/convert_*.py`.

**Comment by OriPekelman** [comment]
**E1.2 landed** (commit `6aca4b7`). `ToyVit.patch_embed(sess, kernel, image, patch_size)` composes conv2d + permute + cont_2d into a single helper. Output shape `ne=[d_model, N_patches]` — same shape as the transformer's sequence input on the LM side.

Bit-identical with PyTorch `nn.functional.conv2d + reshape` to within f32 roundoff (max abs diff = 2.98e-08) on an 8×8×3 image with patch=4.

**E1 progress: 2/6 sub-issues shipped (E1.1 + E1.2).** Next is E1.3 — the ViT forward graph (largest remaining piece: class token + learned pos embed + 12 transformer blocks + final norm + MLP head). E1.4 (timm extractor), E1.5 (image loader), E1.6 (driver) follow.

**Spinel landmine pinned** in commit message and source comment: `ggml_permute(a, axis0, axis1, axis2, axis3)` uses "source axis i goes to result axis args[i]", not "result axis i comes from source axis args[i]". `permute(1,0,2,3)` (the transpose alias) is symmetric so the two conventions collapse — patch_embed needed `permute(1, 2, 0, 3)` to map source-axis-2 (OC) to result-axis-0, not the `permute(2, 0, 1, 3)` you'd write if you assumed the other convention.

**Comment by OriPekelman** [comment]
**E1.3 landed** (commit `8cd1d51`). `ViTTinyForwardFFICache` + `ViTTinyConfig` + `ViTTinyBlockFFI`. End-to-end forward + CE + backward + AdamW on a single-image memorisation smoke (16×16 image, patch=4, d=64, 2L×4H, 10 classes):

```
step  1: loss=2.30258   ← exactly ln(10), uniform-class baseline
step 10: loss=1.56748
VERDICT: ViT training is learning  (ratio 0.68)
```

**E1 progress: 3/6 sub-issues shipped (E1.1, E1.2, E1.3).** Remaining: E1.4 (timm extractor), E1.5 (image loader), E1.6 (driver).

**Surfaced three ggml backward gaps**, filed upstream as [ggml-org/ggml#1514](https://github.com/ggml-org/ggml/issues/1514):
1. `GGML_OP_NORM` (LayerNorm) has no backward — only RMSNorm does
2. `GGML_UNARY_OP_GELU` has no backward — only SiLU among activations does
3. `ggml_conv_2d` ends in `cont(permute(...))` and OP_CONT backward asserts contig grad

Worked around in toy: RMSNorm + SiLU + flat-linear patch_embed (mathematically equivalent to stride=patch / no-overlap conv). The smoke trains end-to-end with the workarounds; timm-loader compat (E1.4) just needs a 3-line conversion in the converter script.

**Comment by OriPekelman** [comment]
**E1 is now 6/6 sub-issues complete** (commits `b022316`, `930ea1b`, `1651b38`, plus E1.1–E1.3 earlier today). Full smoke path:

```
uv run prep/preprocess_images.py
make example_train_vit_tiny
TAO_RUN_DIR=/tmp/vit STEPS=200 ./examples/example_train_vit_tiny
```

200-step memorisation smoke: loss 2.30 → 5.7e-4, ratio 0.000247, quality_gate "passed". Events.jsonl well-formed with run_start (`model.kind="vit"`, full schedule + backend metadata), per-step events, run_end with gate.

**For the *real* Tao E1 protocol** (Flowers-102 / Stanford Cars × 3 inits × 3 seeds × 6000 steps at batch 64), the remaining work is logistics rather than new code:
- preprocess via `MODE=real DATA_DIR=/srv/flowers/train uv run prep/preprocess_images.py`
- run with appropriate config (`IMAGE_SIZE=96 PATCH_SIZE=16 D_MODEL=192 N_LAYERS=12 N_HEADS=3 D_FF=768 NUM_CLASSES=102`)
- warm-start the patch_embed / pos_embed / cls_token from `data/vit_tiny_donor.gguf` via a `tnn_upload` between realize and step 0

**Ggml backward gaps** (ggml-org/ggml#1514) remain — we use RMSNorm + SiLU + flat-linear patch_embed instead of LayerNorm + GELU + Conv2d. timm donor weights load cleanly into the flat-linear convention (E1.4's 3-line reshape).

**Comment by OriPekelman** [comment]
**From Tao dogfooding** (set up a real research home at `~/sites/tao_transfer`, driven by the `tao` tool):

`example_train_vit_tiny` now emits a full `toy/v1` events stream ✅ and runs **from-scratch** through Tao end-to-end — just ran a V1 depth sweep (2 arms × 2 seeds) → reports/conclusion, all green.

The one remaining piece for E1's *reproduction* (granite_transfer #28) is the **warm-init arm**: an `INIT={pure_emb,scratch}` knob in the ViT trainer that, for `pure_emb`, loads the timm `vit_tiny_patch16_224` `patch_embed`/`pos_embed`/`cls_token` (the E1.4 extractor's GGUF) as warm init while the blocks stay random — so **pure-emb-vs-scratch becomes two arms Tao can sweep**. The extractor already shipped; only the training-side load is missing.

Tao's E1 spec is queued in `tao_transfer` and ready to run the moment that knob lands.

**Comment by OriPekelman** [comment]
**INIT knob landed** (commit `53fdc8f`). `examples/07_train_vit_tiny.rb` now honors `INIT={scratch,pure_emb}`:

- `INIT=scratch` (default) — full random init, unchanged.
- `INIT=pure_emb` — loads patch_embed.proj.weight, pos_embed, and cls_token from `data/vit_tiny_donor.gguf` (the E1.4 extractor's output); block weights stay random.

`run_start.config.init` carries the value so Tao can pivot the sweep on it. `config.donor` is the donor GGUF path when `pure_emb`.

**Preflight dim check** bails loud if cfg ≠ donor shape. The donor is fixed at ViT-Tiny native dims (image=224, patch=16, d=192, heads=3, d_ff=768, L=12, seq_t=197), so:

```sh
INIT=pure_emb IMAGE_SIZE=224 PATCH_SIZE=16 D_MODEL=192 N_HEADS=3 D_FF=768 N_LAYERS=12 \\
  ./examples/example_train_vit_tiny
```

Verified: 3 steps, loss 2.30 → 2.14, gate passed, 173 PARAMs tracked, donor load reported. The pure-emb-vs-scratch sweep is now a real two-arm comparison.

**Follow-up (separate concern, not blocking this)**: pos_embed resampling so non-native input sizes (96×96 for Flowers/Cars per the actual granite_transfer #28 protocol) work. timm's `resample_abs_pos_embed` is the textbook recipe; small extractor-side or driver-side script. Filing as part of any next push if you want it.

**Comment by OriPekelman** [comment]
**Shipped — closing.**

All four scope items in the issue are in main:

| Scope | Status | Files |
| --- | --- | --- |
| ViT-Tiny arch (patch_embed, pos_embed, cls_token, 12 blocks, head) | ✅ | `lib/toy_vit.rb`, `lib/vit_tiny_forward_ffi.rb` |
| timm `vit_tiny_patch16_224` (IN-21k AugReg) → GGUF loader | ✅ | `prep/extract_vit_tiny.py` |
| Image-classification training loop + events | ✅ | `examples/07_train_vit_tiny.rb` |
| 200-step acceptance | ✅ | verified just now |

**Acceptance run (just verified, 2026-05-28):**
```
$ STEPS=200 TAO_RUN_DIR=/tmp/vit_acc ./examples/example_train_vit_tiny
initial=2.302585 final=0.000569 ratio=0.000247 gate=passed
$ wc -l /tmp/vit_acc/events.jsonl
202 (run_start + 200 step + run_end, toy/v1 schema)
```

`INIT=pure_emb DONOR_GGUF=...` warm-start path also wired — loads patch_embed + pos_embed + cls_token from the donor GGUF, block weights randomly initialised (the E1 pure-emb arm).

The parenthetical "E1 reproduces the granite_transfer #28 pure-emb result" is downstream Tao work that consumes this. Toy-side is complete.

Closes.
